File: src/convergence_io/dilation.py
"""
dilation.py — Time Dilation Field D(v) for CEG nodes.

D(v) = f(uncertainty, cost_pressure, confidence)

Semantics:
  High uncertainty   → D > 1  (slow region — explore deeply, take more time)
  High confidence    → D < 1  (fast region — execute quickly)
  High cost_pressure → D < 1  (compressed — optimize for speed/cost)

Applied per-node per-tick by the execution loop:
  dilated_latency = base_latency * D(v)
  cost(edge) *= D(source_node)

Range: [D_MIN, D_MAX] — clamped to prevent runaway slowdowns.

Usage:
    from convergence_io.dilation import DilationField, dilation

    d = dilation(uncertainty=0.8, cost_pressure=0.2, confidence=0.5)  # → ~1.4
    field = DilationField()
    field.update_node("node-id", uncertainty=0.3, cost_pressure=0.7, confidence=0.9)
    d = field.get("node-id")
"""

from __future__ import annotations

import math
import threading
from dataclasses import dataclass, field
from typing import Dict, Optional

D_MIN = 0.1    # never compress below 10% of baseline
D_MAX = 5.0    # never dilate above 5× baseline
D_DEFAULT = 1.0


def dilation(
    uncertainty: float,
    cost_pressure: float,
    confidence: float,
) -> float:
    """
    Compute scalar dilation for a single node.

    Args:
        uncertainty:   0.0 (certain) → 1.0 (completely unknown)
        cost_pressure: 0.0 (no pressure) → 1.0 (must minimize cost)
        confidence:    0.0 (no confidence) → 1.0 (fully confident)

    Returns:
        float in [D_MIN, D_MAX]

    Formula:
        raw = (1 + uncertainty) / (1 + confidence) / (1 + cost_pressure)
        Uncertainty inflates dilation.
        Confidence and cost_pressure deflate it.
    """
    uncertainty = max(0.0, min(1.0, uncertainty))
    cost_pressure = max(0.0, min(1.0, cost_pressure))
    confidence = max(0.0, min(1.0, confidence))

    raw = (1.0 + uncertainty) / ((1.0 + confidence) * (1.0 + cost_pressure))
    return max(D_MIN, min(D_MAX, raw))


@dataclass
class NodeDilationState:
    uncertainty: float = 0.5
    cost_pressure: float = 0.0
    confidence: float = 0.5
    value: float = D_DEFAULT


class DilationField:
    """
    Maintains per-node dilation values.
    Updated each tick by the execution loop based on observed runtime signals.
    """

    def __init__(self) -> None:
        self._states: Dict[str, NodeDilationState] = {}
        self._lock = threading.Lock()

    def update_node(
        self,
        node_id: str,
        uncertainty: float = 0.5,
        cost_pressure: float = 0.0,
        confidence: float = 0.5,
    ) -> float:
        """Recompute and store dilation for node_id. Returns new value."""
        d = dilation(uncertainty, cost_pressure, confidence)
        with self._lock:
            self._states[node_id] = NodeDilationState(
                uncertainty=uncertainty,
                cost_pressure=cost_pressure,
                confidence=confidence,
                value=d,
            )
        return d

    def get(self, node_id: str) -> float:
        """Return current dilation for node_id (D_DEFAULT if unknown)."""
        state = self._states.get(node_id)
        return state.value if state else D_DEFAULT

    def update_from_health(self, node_id: str, health: float, latency_ratio: float) -> float:
        """
        Convenience updater: derive dilation from runtime health + latency signals.

        health:        0..1 (1 = fully healthy)
        latency_ratio: observed/target latency (1.0 = on target, >1 = slow)
        """
        uncertainty = 1.0 - health                          # unhealthy → uncertain
        cost_pressure = max(0.0, 1.0 - 1.0 / max(latency_ratio, 0.01))  # slow → pressure
        confidence = health * max(0.0, 1.0 - abs(latency_ratio - 1.0))
        return self.update_node(node_id, uncertainty, cost_pressure, confidence)

    def apply_to_graph(self, graph: Any) -> None:  # type: ignore[type-arg]
        """Write computed dilation values back into graph nodes."""
        for node_id, state in list(self._states.items()):
            node = graph.get_node(node_id)
            if node is not None:
                node.dilation = state.value

    def snapshot(self) -> Dict[str, float]:
        return {nid: s.value for nid, s in self._states.items()}


# ── Swap Convergence (anti-oscillation) ───────────────────────────────────────

class SwapConvergenceGuard:
    """
    Prevents oscillatory provider switching under PCSF + dilation dynamics.

    Tracks recent swap events per node. If the same swap (old→new) occurs
    more than `max_swaps` times within `window_ticks`, the swap is blocked
    and hysteresis is applied (D inflated for the worse candidate).
    """

    def __init__(self, max_swaps: int = 3, window_ticks: int = 10) -> None:
        self.max_swaps = max_swaps
        self.window_ticks = window_ticks
        self._history: Dict[str, list] = {}  # key → [tick, ...]
        self._lock = threading.Lock()

    def record_swap(self, old_id: str, new_id: str, tick: int) -> None:
        key = f"{old_id}→{new_id}"
        with self._lock:
            self._history.setdefault(key, [])
            self._history[key].append(tick)

    def is_oscillating(self, old_id: str, new_id: str, current_tick: int) -> bool:
        key = f"{old_id}→{new_id}"
        with self._lock:
            history = self._history.get(key, [])
            recent = [t for t in history if current_tick - t <= self.window_ticks]
            self._history[key] = recent
            return len(recent) >= self.max_swaps

    def reset(self, old_id: str, new_id: str) -> None:
        key = f"{old_id}→{new_id}"
        with self._lock:
            self._history.pop(key, None)

File: src/cio/src/dilation.rs
/// dilation.rs — Time dilation field D(v)
///
/// D(v) = f(uncertainty, cost_pressure, confidence)
///
/// High uncertainty   → D > 1  (slow region — explore deeply)
/// High confidence    → D < 1  (fast region — execute quickly)
/// High cost_pressure → D < 1  (compressed — minimize cost/time)
///
/// Range: [D_MIN, D_MAX]

pub const D_MIN: f64 = 0.1;
pub const D_MAX: f64 = 5.0;
pub const D_DEFAULT: f64 = 1.0;

/// Compute scalar dilation for one node.
///
/// # Arguments
/// - `uncertainty`   — 0.0 (certain) → 1.0 (unknown)
/// - `cost_pressure` — 0.0 (none) → 1.0 (must minimize)
/// - `confidence`    — 0.0 (no confidence) → 1.0 (fully confident)
///
/// # Formula
/// ```
/// raw = (1 + uncertainty) / ((1 + confidence) * (1 + cost_pressure))
/// ```
pub fn dilation(uncertainty: f64, cost_pressure: f64, confidence: f64) -> f64 {
    let u = uncertainty.clamp(0.0, 1.0);
    let p = cost_pressure.clamp(0.0, 1.0);
    let c = confidence.clamp(0.0, 1.0);
    let raw = (1.0 + u) / ((1.0 + c) * (1.0 + p));
    raw.clamp(D_MIN, D_MAX)
}

/// Per-node dilation state.
#[derive(Debug, Clone)]
pub struct NodeDilation {
    pub uncertainty:   f64,
    pub cost_pressure: f64,
    pub confidence:    f64,
    pub value:         f64,
}

impl NodeDilation {
    pub fn from_health(health: f64, latency_ratio: f64) -> Self {
        let uncertainty   = (1.0 - health).clamp(0.0, 1.0);
        let cost_pressure = (1.0 - 1.0 / latency_ratio.max(0.01)).max(0.0).min(1.0);
        let confidence    = health * (1.0 - (latency_ratio - 1.0).abs()).max(0.0);
        let value = dilation(uncertainty, cost_pressure, confidence);
        NodeDilation { uncertainty, cost_pressure, confidence, value }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn high_uncertainty_inflates() {
        assert!(dilation(0.9, 0.0, 0.5) > dilation(0.1, 0.0, 0.5));
    }

    #[test]
    fn high_confidence_deflates() {
        assert!(dilation(0.5, 0.0, 0.9) < dilation(0.5, 0.0, 0.1));
    }

    #[test]
    fn high_cost_pressure_deflates() {
        assert!(dilation(0.5, 0.9, 0.5) < dilation(0.5, 0.1, 0.5));
    }

    #[test]
    fn clamped_min() {
        assert!(dilation(0.0, 1.0, 1.0) >= D_MIN);
    }

    #[test]
    fn clamped_max() {
        assert!(dilation(1.0, 0.0, 0.0) <= D_MAX);
    }

    #[test]
    fn clamps_out_of_range_inputs() {
        let d = dilation(5.0, -1.0, 3.0);
        assert!((D_MIN..=D_MAX).contains(&d));
    }

    #[test]
    fn from_health_healthy() {
        let nd = NodeDilation::from_health(1.0, 1.0);
        assert!(nd.value <= 1.0);
    }

    #[test]
    fn from_health_degraded() {
        let good = NodeDilation::from_health(1.0, 1.0);
        let bad  = NodeDilation::from_health(0.1, 3.0);
        assert!(bad.value > good.value);
    }
}
